### PR TITLE
io_uring: add io_uring socket handle

### PR DESCRIPTION
**Pull request by zhxie**
Signed-off-by: He Jie Xu <hejie.xu@intel.com>
Signed-off-by: Xie Zhihao <zhihao.xie@intel.com>

Commit Message: io_uring: add io_uring socket handle
Additional Description: This is a pretty large (mostly in tests) patch which introduces `IoUringSocketHandle`. `IoUringSocketHandle` benefits from `IoUringWorker`, `IoUringWorkerFactory` and `IoUringSocket` and exposes io_uring features to the upper components within the `IoHandle` API. `IoUringSocketHandle` does only support UDP-related features. In the next patch, we will define the API and merge io_uring into the socket interface to make io_uring work in Envoy.
Risk Level: Low
Testing: Unit and integration tests
Docs Changes: N/A (will be added when the socket interface and its API are ready) 
Release Notes: N/A
Platform Specific Features: N/A
Part of #28395
Fix #32265 #32946

### CODE REVIEW

**Comment by zhxie** [comment]
It is the third attempt of adding io_uring socket handle. So I would like to conclude the status and list the test plan.

| CI target                                	| Reason                                                                                 	| Solution                                                                                                                                                                                                         	|
|------------------------------------------	|----------------------------------------------------------------------------------------	|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------	|
| mobile/android (#32265, e8ea33a, #32936) 	| Android also build with \_\_LINUX\_\_ defined. PR does not trigger mobile CI.          	| Avoid building with \_\_ANDROID_API\_\_. Then Touch mobile/README.md to trigger the CI.                                                                                                                          	|
| mobile/perf (#32946, 45fa977, #33090)    	| As suggested, ensure io_uring_socket_handle_impl is not built to limit size of mobile. 	| Change mobile-perf.yml and remove io_uring_socket_handle_impl.cc.                                                                                                                                                	|
|                                          	| Change CI YAML does not take effect until merging.                                     	| Revert the previous change and exclude io_uring_socket_handle_impl by //bazel:android in BUILD file. Then add unbuildable code for Android in io_uring_socket_handle_impl to ensure the file would not be built. 	|

**Comment by yanavlasov** [comment]
Please merge main to resolve build errors.

/wait

**Review by soulxu — CHANGES_REQUESTED** [review]

File: mobile/README.md
Comment (zhxie) [reply mobile/README.md:43]: True. I will revert 4de1200 and df76561 after approval and before merging.

**Comment by yanavlasov** [comment]
Thanks for making this change. Started to review this. Will finish today or tomorrow.

**Comment by RyanTheOptimist** [comment]
@yanavlasov - friendly maintainer-on-call ping

**Comment by adisuissa** [comment]
@KBaichoo @yanavlasov PTAL.

File: source/common/network/io_uring_socket_handle_impl.h
Comment (KBaichoo) [inline source/common/network/io_uring_socket_handle_impl.h:12]: nice, so effectively if this somehow leaks into mobile we get a compile-time error, correct?

File: source/common/network/BUILD
Comment (KBaichoo) [inline source/common/network/BUILD:266]: Any idea why we don't need this for ios? e.g. `bazel:ios_arm64`

**Review by KBaichoo — COMMENTED** [review]
Thank you and sorry for the delay

File: source/common/network/io_uring_socket_handle_impl.h
Comment (zhxie) [reply source/common/network/io_uring_socket_handle_impl.h:12]: Yes, we don't want io_uring socket handle in Android.

File: source/common/network/BUILD
Comment (zhxie) [reply source/common/network/BUILD:266]: Since the iOS CI does not complain about io_uring, I guess that iOS does not build //bazel:linux targets but Android does.

**Review by KBaichoo — DISMISSED** [review]

**Comment by zhxie** [comment]
@yanavlasov kindly remind.

File: source/common/network/io_uring_socket_handle_impl.cc
Comment (yanavlasov) [inline source/common/network/io_uring_socket_handle_impl.cc:140]: For my education, are sendmsg/recvmsg not supported for io_uring?

**Review by yanavlasov — DISMISSED** [review]
LGTM. I have left a question about some of the code.

File: source/common/network/io_uring_socket_handle_impl.cc
Comment (zhxie) [reply source/common/network/io_uring_socket_handle_impl.cc:140]: Both sendmsg and recvmsg are supported by io_uring. However, at this stage we only implement TCP-related methods, and we will also determine whether the socket is TCP or not to decide whether to enable io_uring in the socket interface patch later.

**Comment by zhxie** [comment]
@soulxu Since maintainers have approved this patch, I have removed those test-only commits, and I think we can start the merge process then.

**Review by soulxu — APPROVED** [review]

**Pull request merged**
